Adding more fields to Artifact

hnpl opened the pull request:
The newly added fields should be useful for identifying a desired
Artifact from a list of Artifact objects.

Signed-off-by: Hoa Nguyen <hoanguyen@ucdavis.edu>

powerjg reviewed (commented):
Seems fine to me, but it looks like you need to update the tests.

hnpl commented:
Do you think that using Enum's (instead of str) for gem5 versions to avoid variations of a single type (like "v20-1" and "v20.1" for gem5 20.1)?

Also, should the benchmark name be part of the Artifact?

powerjg commented on artifact/gem5art/artifact/artifact.py:
```diff
@@ -100,6 +100,18 @@ class Artifact:
+
+    Optional fields:
+    a) architecture: name of the ISA (e.g. x86, riscv) ("" by default)
+    b) size: size of the artifact in bytes (-1 by default)
```
Probably should be None to denote N/A instead of -1

hnpl committed:
Adding more fields to Artifact

The newly added fields should be useful for identifying a desired
Artifact from a list of Artifact objects.

Signed-off-by: Hoa Nguyen <hoanguyen@ucdavis.edu>

powerjg commented:
Version shouldn't be an enum. We don't want to have to update it on every update to gem5. 

I'd vote against benchmark for all artifacts.

However, it may be nice to be able extend an artifact with specific fields. There's no reason we can't do that with json/mongo/etc. Maybe we should also allow `**kwargs` and have a dictionary with "extra" fields that we save to the database/json/etc.

hnpl committed:
Fix typos

Signed-off-by: Hoa Nguyen <hoanguyen@ucdavis.edu>

hnpl commented:
I opted to flatten the Artifact.extra to Artifact fields (i.e. Artifact.extra_field_1, Artifact.extra_field_2, etc.)

hnpl commented:
Though, I have no reason other than we won't need to re-flatten the "extra" field again while querying for artifacts

hnpl committed:
Update

Signed-off-by: Hoa Nguyen <hoanguyen@ucdavis.edu>

powerjg commented on artifact/gem5art/artifact/artifact.py:
```diff
@@ -168,6 +197,19 @@ def registerArtifact(cls,
+
+        for k, v in kwargs.items():
+            if k in data:
+                raise Exception("Field `{}` is redefined".format(k))
```
This can't be, right? Because `**kwargs` is whatever is left over.

I think rather than saying that these extra things have a __str__, you should restrict them to be *only strings*. I.e., the type should be `**kwargs: str`. See https://www.python.org/dev/peps/pep-0484/#arbitrary-argument-lists-and-default-argument-values

hnpl replied:
Yeah, it would be a TypeError if a keyword arg is redefined.

hnpl committed:
Update kwargs

Signed-off-by: Hoa Nguyen <hoanguyen@ucdavis.edu>

hnpl commented:
Ah, I just remember why I checked the keywords. There are fields in data that are not part of the function arguments (i.e. "hash"). So I'd prefer to check the keywords before adding to the data dictionary, or have the extra field being a dictionary.

hnpl committed:
Roll back key checking

Signed-off-by: Hoa Nguyen <hoanguyen@ucdavis.edu>
